sphinx/builders/__init__.py

```diff
@@ -403,6 +405,109 @@ def build(self, docnames, summary=None, method='update'):
         self.finish_tasks.join()
 
+    def read(self):
+        # type: () -> List[unicode]
+        """(Re-)read all files new or changed since last update.
+
+        Store all environment docnames in the canonical format (ie using SEP as
+        a separator in place of os.path.sep).
+        """
+        changed, reason = self.env.update_config(self.config, self.srcdir, self.doctreedir)
+
+        # this cache also needs to be updated every time
+        self.env._nitpick_ignore = set(self.config.nitpick_ignore)
+
+        logger.info(bold('updating environment: '), nonl=True)
+
+        self.env.find_files(self.config, self)
+        added, changed, removed = self.env.get_outdated_files(changed)
+
+        # allow user intervention as well
+        for docs in self.app.emit('env-get-outdated', self, added, changed, removed):
+            changed.update(set(docs) & self.env.found_docs)
+
+        # if files were added or removed, all documents with globbed toctrees
+        # must be reread
+        if added or removed:
+            # ... but not those that already were removed
+            changed.update(self.env.glob_toctrees & self.env.found_docs)
+
+        if changed:
+            logger.info('[%s] ', reason, nonl=True)
+        logger.info('%s added, %s changed, %s removed',
+                    len(added), len(changed), len(removed))
+
+        # clear all files no longer present
+        for docname in removed:
+            self.app.emit('env-purge-doc', self.env, docname)
+            self.env.clear_doc(docname)
+
+        # read all new and changed files
+        docnames = sorted(added | changed)
+        # allow changing and reordering the list of docs to read
+        self.app.emit('env-before-read-docs', self.env, docnames)
+
+        # check if we should do parallel or serial read
+        if parallel_available and len(docnames) > 5 and self.app.parallel > 1:
+            par_ok = self.app.is_parallel_allowed('read')
+        else:
+            par_ok = False
+
+        if par_ok:
+            self._read_parallel(docnames, nproc=self.app.parallel)
+        else:
+            self._read_serial(docnames)
+
+        if self.config.master_doc not in self.env.all_docs:
+            raise SphinxError('master file %s not found' %
+                              self.env.doc2path(self.config.master_doc))
+
+        for retval in self.app.emit('env-updated', self.env):
+            if retval is not None:
+                docnames.extend(retval)
+
+        return sorted(docnames)
+
+    def _read_serial(self, docnames):
+        # type: (List[unicode]) -> None
+        for docname in status_iterator(docnames, 'reading sources... ', "purple",
+                                       len(docnames), self.app.verbosity):
+            # remove all inventory entries for that file
+            self.app.emit('env-purge-doc', self.env, docname)
+            self.env.clear_doc(docname)
+            self.env.read_doc(docname, self.app)
```

Comment:
As I commented first, I will work about `env.read_doc()` in next PR. Stay tuned :-)